text/0000-infallible-promotion.md

```diff
@@ -0,0 +1,193 @@
+* All code handling constants has to carry [extra complexity to support promotion](https://github.com/rust-lang/rust/issues/75461)
+
+This RFC proposes to fix all these problems at once, by restricting implicit promotion to those expression whose evaluation cannot fail.
+This is the last step in a series of changes that have been going on for quite some time, starting with the [introduction](https://github.com/rust-lang/rust/pull/53851) of the `#[rustc_promotable]` attribute to control which function calls may be subject to implicit promotion (the original RFC said that all calls to `const fn` should be promoted, but as user-defined `const fn` got closer and closer, that seemed less and less like a good idea, due to all the ways in which evaluating a `const fn` can fail).
+Together with [some planned changes for evaluation of regular constants](https://github.com/rust-lang/rust/issues/71800), this means that all CTFE failures can be made hard errors, greatly simplifying the parts of the compiler that trigger evaluation of constants and handle the resulting value or error.
+
+For more details, see [the MCP that preceded this RFC](https://github.com/rust-lang/lang-team/issues/58).
+
+[promotion-status]: https://github.com/rust-lang/const-eval/blob/33053bb2c9a0c6a17acd3116dd47bbb360e060db/promotion.md
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+(Based on [RFC 1414][rfc1414])
+
+Inside a function body's block:
+
+- If a shared reference to a constexpr rvalue is taken. (`&<constexpr>`),
+- And the constexpr does not contain a `UnsafeCell { ... }` constructor,
+- And the constexpr only consists of operations that will definitely succeed to
+  evaluate at compile-time,
+- And the resulting value does not need dropping,
+- Then instead of translating the value into a stack slot, translate
+  it into a static memory location and give the resulting reference a
+  `'static` lifetime.
+
+Operations that definitely succeed include:
+- literals of any kind
+- constructors (struct/enum/union/tuple)
+- struct/tuple field accesses
+- arithmetic that does not involve division: `+`/`-`/`*`
+
+Note that arithmetic overflow is not a problem: an addition in debug mode is compiled to a `CheckedAdd` MIR operation that never fails, which returns an `(<int>, bool)`, and is followed by a check of said `bool` to possibly raise a panic. We only ever promote the `CheckedAdd`, so evaluation of the promoted will never fail, even if the operation overflows. For example, `&(1 + u32::MAX)` turns into something like:
+```rust
+const C: (u32, bool) = CheckedAdd(1, u32::MAX); // evaluates to (1, true).
+assert!(C.1 == false);
+&C.0
+```
+
+Operations that might fail include:
+- `/`/`%`
+- `panic!` (including the assertion that follows `Checked*` arithmetic to ensure that no overflow happened)
+- array/slice indexing
+- any unsafe operation
+- `const fn` calls (as they might do any of the above)
+
```

Comment:
I am pretty sure `<<` and `>>` are no arithmetic operation, so it would need clarification.

Reply:
The RFC states that infallibe operations can be promoted. `<<` and `>>` are infallible (they are treated the same as `+`). Hence they can be promoted. The following list just gives some examples.

But it seems not making that list exhaustive is impossible so sure, I'll just add those as well.